scipy/linalg/_decomp_update.pyx

```diff
@@ -0,0 +1,2416 @@
+    # i.e. u -= q.dot(s)
+    if qisF:
+        gemv(N, m, n, -1, q, qs[1], s, 1, 1, u, 1)
+    else:
+        gemv(T, n, m, -1, q, n, s, 1, 1, u, 1)
+    wnorm = nrm2(m, u, 1)
+
+    if blas_t_less_than(inv_root2, wnorm):
+        scal(m, 1/wnorm, u, 1)
+        s[n] = wnorm
+        return True
+
+    # if the above check failed, try one reorthogonalization
+    if qisF:
+        if blas_t is float or blas_t is double:
+            gemv(T, m, n, 1, q, qs[1], u, 1, 0, s+n, 1)
+        else:
+            gemv(C, m, n, 1, q, qs[1], u, 1, 0, s+n, 1)
+        gemv(N, m, n, -1, q, qs[1], s+n, 1, 1, u, 1)
+    else:
+        if blas_t is float or blas_t is double:
+            gemv(N, n, m, 1, q, n, u, 1, 0, s+n, 1)
+        else:
+            blas_t_conj(m, u, &ss)
+            gemv(N, n, m, 1, q, n, u, 1, 0, s+n, 1)
+            blas_t_conj(m, u, &ss)
+            blas_t_conj(n, s+n, &ss)
+        gemv(T, n, m, -1, q, n, s+n, 1, 1, u, 1)
+        
+    wpnorm = nrm2(m, u, 1) 
+    
+    if blas_t_less_than(wpnorm, wnorm/inv_root2): # u lies in span(q) 
+        scal(m, 0, u, 1)
+        axpy(n, 1, s, 1, s+n, 1)
+        s[n] = 0
+        return False
+    scal(m, 1/wpnorm, u, 1)
+    axpy(n, 1, s, 1, s+n, 1)
+    s[n] = wpnorm
+    return True
+
+cdef int thin_qr_row_delete(int m, int n, blas_t* q, int* qs, bint qisF, blas_t* r, int* rs, int k, int p_eco, int p_full) nogil:
+    cdef int i, j, argmin_row_norm 
+    cdef size_t usize = (m + 3*n + 1) * sizeof(blas_t)
+    cdef blas_t* s
+    cdef blas_t* u
+    cdef blas_t* s1
+    cdef int us[1]
+    cdef int ss[1]
+    cdef blas_t c, sn, min_row_norm, row_norm
+
+    u = <blas_t*>libc.stdlib.malloc(usize)
+    if not u:
+        return MEMORY_ERROR
+    s = u + m
+    ss[0] = 1
+    ss[1] = 0
+    us[0] = 1
+    us[1] = 0
+
+    for i in range(p_eco):
+        memset(u, 0, usize)
+        # permute q such that row k is the last row.
+        if k != m-1:
+            for j in range(k, m-1):
+                swap(n, row(q, qs, j), qs[1], row(q, qs, j+1), qs[1])
+
+        if not reorthx(m, n, q, qs, qisF, m-1, u, s):
+            # if we get here it means that this basis vector lies in span(q).
+            # we want to use s[:n+1] but we need a vector into null(q)
+            # find the row of q with the smallest norm and try that. (Daniel, p785)
+            min_row_norm = nrm2(n, row(q, qs, 0), qs[1])
+            argmin_row_norm = 0
+            for j in range(1, m):
+                row_norm = nrm2(n, row(q, qs, j), qs[1])
+                if blas_t_less_than(row_norm, min_row_norm):
+                    min_row_norm = row_norm
+                    argmin_row_norm = j
+            memset(u, 0, m*sizeof(blas_t))
+            if not reorthx(m, n, q, qs, qisF, argmin_row_norm, u, s+n+1):
+                # failed, quit.
+                libc.stdlib.free(u)
+                return 0
+
+        memset(s+2*n, 0, n*sizeof(blas_t))
+
+        # what happens here...
```

Comment:
It is mostly a point where I don't really know what the right answer is.  When updating an economic mode decomposition, a vector into the null space of Q (equivalently A, if A = QR) is needed.  The papers I've been working from either do, or suggest doing what I do here, but its probably possible to construct an example where this will fail.  It probably needs to have some kind of "try even harder" fallback.  Alternatively, it might never happen that this fails, I'm not really sure yet.